src/vm/dllimport.cpp

```diff
@@ -5913,37 +5913,47 @@ static HMODULE LocalLoadLibraryHelper( LPCWSTR name, DWORD flags, LoadLibErrorTr
 
     HMODULE hmod = NULL;
 
+    AppDomain* pDomain = GetAppDomain();
+
+    hmod = pDomain->FindUnmanagedImageInCache(name);
+
+    if (hmod == NULL)
+    {
 #ifndef FEATURE_PAL
 
     if ((flags & 0xFFFFFF00) != 0
 #ifndef FEATURE_CORESYSTEM
         && NDirect::SecureLoadLibrarySupported()
 #endif // !FEATURE_CORESYSTEM
-        )
-    {
-        hmod = CLRLoadLibraryEx(name, NULL, flags & 0xFFFFFF00);
-        if(hmod != NULL)
+            )
         {
-            return hmod;
-        }
+            hmod = CLRLoadLibraryEx(name, NULL, flags & 0xFFFFFF00);
+            if (hmod != NULL)
+            {
+                pDomain->AddUnmanagedImageToCache(name, hmod);
+                return hmod;
+            }
 
-        DWORD dwLastError = GetLastError();
-        if (dwLastError != ERROR_INVALID_PARAMETER)
-        {
-            pErrorTracker->TrackErrorCode();
-            return hmod;
+            DWORD dwLastError = GetLastError();
+            if (dwLastError != ERROR_INVALID_PARAMETER)
+            {
+                pErrorTracker->TrackErrorCode();
+                return hmod;
+            }
         }
-    }
 
     hmod = CLRLoadLibraryEx(name, NULL, flags & 0xFF);
-    
+
 #else // !FEATURE_PAL
     hmod = CLRLoadLibrary(name);
 #endif // !FEATURE_PAL
-        
-    if (hmod == NULL)
-    {
-        pErrorTracker->TrackErrorCode();
+
+        pDomain->AddUnmanagedImageToCache(name, hmod);
```

Comment:
Why would we cache the failure in this case? I would assume or rather assert that the cache should never have a null value for a key. Is there a scenario where a null module would be acceptable?